observer/src/filters/vision/ball/BallAssigner.cpp

```diff
@@ -0,0 +1,114 @@
+//
+// Created by rolf on 08-07-21.
+//
+
+#include "observer/filters/vision/ball/BallAssigner.h"
+#include <algorithm>
+namespace BallAssigner{
+BallObservation mergeObservationsByArea(const std::vector<BallObservation>& observations){
+  //It's the user's responsibility to ensure all observations are from the same detectionFrame;
+  //otherwise, merging makes very little sense
+  assert(!observations.empty());
+#ifndef NDEBUG
+  for(const auto& observation : observations){
+    assert(observation.cameraID == observations[0].cameraID);
+    assert(observation.timeCaptured == observations[0].timeCaptured);
+    assert(observation.timeSent == observations[0].timeSent);
+  }
+#endif
+  unsigned int totalArea = 0;
+  for(const auto& observation : observations){
+    totalArea+=observation.area;
+  }
+  bool zero_area = totalArea <= 0;
+  Eigen::Vector2d position(0,0);
+  Eigen::Vector2d pixelPosition(0,0);
+  double confidence = 0.0;
+  double height = 0.0;
+  for(const auto& observation : observations){
+    double weight = zero_area ? 1.0/static_cast<double>(observations.size()) :
+        observation.area / static_cast<double>(totalArea);
+    position += weight * observation.position;
+    pixelPosition += weight * observation.pixelPosition;
+    confidence = fmax(confidence,observation.confidence);
+    height += weight * observation.height;
+  }
+  return BallObservation(observations[0].cameraID,observations[0].timeCaptured,observations[0].timeSent,
+                         position,pixelPosition,confidence,totalArea,height);
+}
+
+
+BallAssignmentResult assign_balls(const std::vector<CameraGroundBallPrediction>& predictions,
+                                  const std::vector<BallObservation>& observations) {
+  BallAssignmentResult assignment(predictions);
+  std::vector<std::vector<BallObservation>> prediction_observations(predictions.size());
+
+  static constexpr double MAX_ACCEPT_DIST = 0.3;
+  //first pass: greedily assign observations to closest predictions
+  //this will probably become more complicated in the future as it will work on a per-object basis
+  for(const auto& observation : observations){
+      auto closest_prediction = std::min_element(assignment.op_pairs.begin(),assignment.op_pairs.end(),
+                                                 [&](const CameraGroundBallPredictionObservationPair& first,
+                                                     const CameraGroundBallPredictionObservationPair& second){
+                                                   return (observation.position - first.prediction.position).squaredNorm() <
+                                                       (observation.position - second.prediction.position).squaredNorm();
+                                                 });
+      //if there's no close ball, we will use it for new filters.
+      if(!assignment.op_pairs.empty() ||
```

Comment:
If op_pairs is empty, closest_prediction will return an invalid iterator. closest_prediction->prediction.position-observation.position in L58 will therefore cause a segfault. Was the || supposed to be a &&?

Reply:
Silly mistake: The not sign was not supposed to be there. Good catch.